lib/actions/usbsdfuTargetActions.js

```diff
@@ -268,103 +287,196 @@ function calculateSHA512Hash(image) {
+        sdReq = userInput.sdSdReq;
+        fwVersion = userInput.sdVer;
+    } else if (image.name.includes(RegionName.APPLICATION)) {
+        sdReq = userInput.appSdReq;
+        fwVersion = userInput.appVer;
+    } else {
+        throw new Error('Firmware type is unknown when setting user input.');
+    }
+
+    image = sdReq ? image.setIn(['initPacket', 'sdReq'], sdReq) : image;
+    image = fwVersion ? image.setIn(['initPacket', 'fwVersion'], fwVersion) : image;
+    image = userInput.isDebug ? image.setIn(['initPacket', 'isDebug'], userInput.isDebug) : image;
+
+    return image;
+}
+
+// Update calculated hash regarding to the DFU image
+function handleHash(image, hashType) {
+    let hash;
+    switch (hashType) {
+        case initPacket.HashType.NO_HASH:
+            logger.error('No hash is not supported');
+            break;
+        case initPacket.HashType.SHA128:
+            logger.error('Hash type SHA128 is not supported');
+            break;
+        case initPacket.HashType.SHA256:
+            hash = calculateSHA256Hash(image.firmwareImage);
+            logger.info('Hash is generated by SHA256');
+            break;
+        case initPacket.HashType.SHA512:
+            hash = calculateSHA512Hash(image.firmwareImage);
+            logger.info('Hash is generated by SHA512');
+            break;
+        case initPacket.HashType.CRC:
+            logger.error('Hash type CRC is not supported');
+            break;
+        default:
+            logger.error('Unknown hash type');
+    }
+
+    return image.setIn(['initPacket', 'hashType'], hashType)
+                .setIn(['initPacket', 'hash'], hash);
+}
 
-        dispatch(targetActions.writeProgressStartAction());
-        dfu.start(true)
+// Operate DFU process with update all the images on the target device
+function operateDFU(inputDfuImages) {
+    if (inputDfuImages.size === 0) {
+        logger.info('All dfu images have been written to the target device.');
+        return dispatch => Promise.resolve()
             .then(() => {
-                logger.info('DFU completed successfully!');
                 dispatch(targetActions.writeProgressFinishedAction());
-                dispatch({ type: 'DEVICE_DESELECTED' });
-            })
-            .catch(error => {
-                logger.error(`DFU failed. Reason: ${error.message || error}`);
-            })
-            .then(() => {
-                port.close();
-                logger.info(`Port is closed: ${port.path}`);
+                dispatch(canWrite());
             });
+    }
+
+    const dfuImages = inputDfuImages;
+    let image;
+    let serialNumber;
+
+    return async (dispatch, getState) => Promise.resolve()
+        // When dfu starts
+        .then(async () => {
+            serialNumber = getState().app.target.serialNumber;
+            image = dfuImages.first();
+            const packet = await initPacket.createInitPacketUint8Array(image.initPacket);
```

Comment:
`initPacket` functions are all synchronous since the loading of protobuf from file is eliminated, so async/await is not needed. (I think the jsdoc in initPacket needs to be updated).

Reply:
Yes, the initPacket implementation has been updated, await and async can be removed.